Python bindings - complete implementation

franekp opened the pull request:
Good place to start: https://github.com/franekp/libcypher-parser/blob/pycypher-wip/pycypher/docs/DeveloperGuide.md

Readme: https://github.com/franekp/libcypher-parser/tree/pycypher-wip/pycypher

cleishm commented on pycypher/README.md:
```diff
@@ -0,0 +1,46 @@
+
+### Getting started
+You don't need to have libcypher-parser installed to use it.
+The distribution contains a statically linked copy of libcypher-parser. You
```
Needs updating if it's integrated with the source tree?

franekp replied:
What do you mean?

franekp replied:
Ok, I think I understand what you mean. The word "distribution" referes here to Python wheels which are generated by build. This is not stored in the source tree.

I think static linking is important so that users will not have to worry about compatiblity between version of the library and the bindings, and will not have to install the library system-wide to use it from Python.

cleishm replied:
The source distribution contains only the code, not a statically linked copy of the library. If this file is only relevant in the binary distribution, it might be good to change the name in the source tree so that the casual reader isn't assuming it's about the source.

cleishm replied:
(Wrote that comment before seeing your last one. Seems fine to me, with the suggestion to rename the file remaining.)

franekp replied:
Ok, I will change the wording here so this will be more clear. Right now there are no source distributions of the bindings on PyPI and I think this will not change.

cleishm replied:
Great. You could also just rename the file, so that it doesn't look like a readme intended for whomever downloads the source tree.

franekp replied:
Well, a Readme is not only for developers downloading the source tree, right?
For example, here: https://github.com/cleishm/libcypher-parser/blob/master/README.md
there are instructions for end users on how to install from Homebrew or PPA and how to use it.
So I did something similar, treating the Readme as a homepage of the project. I made PyPI page link to this readme (well, the link will become valid after merging this PR).

cleishm replied:
Ok, sounds fine. Just adjust the language then, so that it doesn't sound like there should be a statically linked library bundled with the source (e.g. "The PyCypher packages include a statically linked version of libcypher-parser, so there is no need to install it separately").

franekp replied:
Done, I changed the wording.

cleishm commented on pycypher/local_tests.sh:
```diff
@@ -0,0 +1,23 @@
+export C_INCLUDE_PATH=../src/lib/:$C_INCLUDE_PATH
+
+./generate.py
+./setup.py build_ext -i
```
Why separate scripts for setup and generate?

franekp replied:
Every Python package needs to have a setup.py in it's root, it's a standard. generate.py is separate because it depends on files from parent directory and setup.py is used by Python packaging tools so it's better it does not make assumptions on it's location in the filesystem.

cleishm replied:
Ok. Perhaps generate.py should then be run automatically by configure, or otherwise included into it?

franekp replied:
This sounds good

franekp replied:
Sounds good, I will look into that.

franekp replied:
After a bit of thinking I concluded that pycypher will not be included in `make dist` so I think this no longer apply. The `docker-pycypher` target invokes it, and so the `make check` as part of testing script. `generate.py` runs instantly so it is not a problem that it will be re-run each time `make check` is done.

cleishm commented on pycypher/generate.py:
```diff
@@ -0,0 +1,530 @@
+#!/usr/bin/env python
+# Copyright 2017, Google Inc. All rights reserved.
```
Google?

cleishm commented on pycypher/osx_build.sh:
```diff
@@ -0,0 +1,55 @@
+# limitations under the License.
+set -e -x
+
+if [ "$TRAVIS" != "true" ] || [ "$CI" != "true" ]; then
```
Perhaps also rename the script or move it to a "travis" or "ci" directory?

franekp replied:
Ok, I can rename the whole `utils/` folder to `ci/`

cleishm replied:
Better might be to create a folder at the root, as these scripts could be used for building all packages through the CI (not just python).

cleishm commented on pycypher/local_tests.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+# Copyright 2017, Google Inc. All rights reserved.
```
Google?

franekp replied:
Yes, this is on company time. All the paperwork is already set up, no need to worry.

cleishm replied:
Ok. Perhaps create an AUTHORS file in the root to give context?

franekp replied:
I don't think this is necessary. I think that the notices on top of files clearly specify copyright holders and `git log` is enough for attribution purposes.

cleishm replied:
Ok. Speaking of paperwork, I should probably get a CLA agreement signed to make sure there's no copyright holder issues later. I do that already for libneo4j-parser, but I haven't done so for this project yet (as I was the only contributor until now).

Here's the CLA link: https://www.clahub.com/agreements/cleishm/libcypher-parser

Does this work for you/google?

franekp replied:
Well, good to know. I assumed that there is no CLA for this project since it is not mentioned anywhere in the Readme and project page. Currently I have an approval for patching a project without a CLA.

cleishm replied:
I just created the CLA, since there wasn't one. Check if you have approval for that, and we can chat if otherwise.

cleishm replied:
(To be clear, I'm open to being convinced we don't need one for this project!)

franekp replied:
I applied for approval, we will see what will happen.

cleishm replied:
Hey @franekp. How did that approval go? I did some research of my own, and I also think we could skip it for this sort of project. The main concern needing a CLA is ensuring that it's easy to adjust licensing and copyright later. In this case, I don't see that being as big an issue. What do you think?

franekp replied:
I got an approval, so I can sign the CLA.

cleishm commented on pycypher/local_tests.sh:
```diff
@@ -0,0 +1,23 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Should this file perhaps be a Makefile / Makefile.am?

franekp replied:
This is only for testing, this will not generate build artifacts suitable for uploading. Also, the preferred way is to use linux_build.sh which runs tests on various versions of Python and on 32 and 64 bit arch, it also generates packages ready for upload.

I would leave these two scripts as-is and call one of them from a Makefile so that this can be switched easily.

cleishm replied:
Ok, but having it in a Makefile means the include paths, etc, are already set up. Is "linux_build.sh" and "local_test.sh" a standard in Python packaging? (Apologies for my lack of familiarity). If not, perhaps they can all go into a Makefile? I dislike having arbitrary build scripts lying around when make is really designed for this purpose.

franekp replied:
No, these are custom scripts. I can put them as commands in a Makefile to be consistent with the rest of the project.

franekp replied:
I moved the `local_tests.sh` to `tests/` and added it to `TESTS` and to `dist_check_SCRIPTS` in `Makefile.am`. I moved other build scripts to `<repository-root>/ci` folder.

franekp commented on pycypher/LICENSE:
```diff
@@ -0,0 +1,177 @@
+
```
This is to include the license in built distributions (which apparently also contain source code), just to be safe.

cleishm replied:
Ok. Perhaps we can copy it in at build time? I like to avoid the duplication if it's not necessary.

franekp replied:
It is not necessary but it definitely is not harmful here, the text of the license will not change. Secondly, it makes it more explicit, otherwise users may wonder whether the license from the root directory applies.
If you think that there is a good reason for adding yet another complexity to the build, I can change it so that it would be copied at build time.

cleishm replied:
I just don't like the duplication. None of the other directories in the source tree contain a license file - it should be assumed that the project license applies to all the source.

franekp replied:
Ok, I will add copying the license to the build scripts.

franekp replied:
Done.

franekp commented on .dockerignore:
```diff
@@ -0,0 +1,51 @@
+**/.DS_Store/**/*
```
The Linux build is based on Docker.

cleishm replied:
Is it possible to use the package build by `make dist` rather than copying? I do this for running docker builds for local testing (see [Makefile.am](https://github.com/cleishm/libcypher-parser/blob/master/Makefile.am#L28-L36)).

franekp replied:
No. The manylinux1 standard has restrictions on versions of symbols from glibc that are used so the c library has to be built inside docker to conform to these.

What is generated by `make dist`?

franekp replied:
Also, what do you mean by `source packaging for distribution`?

cleishm replied:
The source tree from github contains the minimal source, and requires extra tools to get setup (autoconf, peg, etc).

Running `make dist` produces a source package that can be built on any system with minimal dependencies. I.e. it contains the generated configure script, the outputs from leg, etc.

This is the package that gets distributed in GitHub releases, and is the basis for all the various OS packaging (homebrew, debian, ubuntu, etc).

franekp replied:
Ok, I think I see your point. This is not a concern when linking statically, right?

cleishm replied:
This isn't related to runtime or static linking. This is just the process by which distribution and packaging occur in autoconf projects. Basically it goes:

* Start with source checked out of version control (github)
* `autogen.sh && configure && make dist` => libcypher-parser-X.X.X.tar.gz
* Copy tarball to target environment: `tar xzf libcypher-parser-X.X.X.tar.gz && cd libcypher-parser-X.X.X && ./configure && make && make install` (or whatever is appropriate for packaging in that environment).

franekp replied:
Ok. Thanks for explanation.

franekp replied:
Done, the build in docker now uses source distribution.

franekp commented on configure.ac:
```diff
@@ -29,7 +29,8 @@ GCC_CFLAGS="-fvisibility=hidden -pipe -Wall -W -Werror \
     -Wno-error=unused-variable -Wno-error=deprecated-declarations"
 
-LT_INIT
+AC_PROG_LIBTOOL
```
Oh, this was by a mistake. I can fix it up.

Linux build needs it because it runs on some special environment to make the shared libraries portable between linux distiributions (more info: https://www.python.org/dev/peps/pep-0513/).

So lines are changed (now the change happens in docker and not affects the working directory directly) before the linux build starts (see dockerfiles in `utils/`).

cleishm replied:
I don't see any mention of libtool in that link? What happens if you try to use LT_INIT instead of AC_PROG_LIBTOOL? I'd have assumed they'd behave identically.

franekp replied:
So this uses some very old CentOS 5 and LT_INIT simply didn't work there.

cleishm replied:
Ok. But referring to my comment regarding the output of `make dist` - if you generate the source distribution first, and then copy that to the docker environment for building the python packaging, you won't need to run autoreconf, etc. The package will already contain everything needed for libtool, etc, to work and it shouldn't matter if the system is older.

franekp replied:
Ok, this sounds good.

franekp replied:
Fixed.

cleishm commented on .travis.yml:
```diff
@@ -0,0 +1,27 @@
+
+deploy:
+  provider: script
+  script: pycypher/utils/upload_wheels.sh $PYPI_USER $PYPI_PASSWORD
```
Also, how would signing of artifacts work? The signing needs to be done locally, where I can access my GPG key.

franekp replied:
Python tools unfortunately don't verify downloaded packages, even if you sign them. Right now uploaded packages are not signed.

cleishm replied:
Ok. I do sign the source packages published to GitHub releases. I'll have to think about a good way to do that now that they're built by CI.

franekp commented:
I addressed all the issues, please reexamine the PR. Sorry for the long delay.
As before, for information about how building and generation works see https://github.com/franekp/libcypher-parser/blob/pycypher-wip/pycypher/docs/DeveloperGuide.md

cleishm commented:
Hi @franekp,

I did some shuffling in the main branch, so I've rebased this PR for you.

I'd also like to see it a little more integrated with autotools - ideally so that a `./configure --with-python && make all` will build the python bindings as well (and `make check` will test them).

I added a commit (e59e656) that starts that integration, but it's not completed. I'm looking at posts like this for techniques: https://blog.kevin-brown.com/programming/2014/09/24/combining-autotools-and-setuptools.html

I'm not, however, familiar enough with python packaging and the use of setuptools. Perhaps you could take a look and provide some thoughts or add some code?

franekp commented:
Hi Chris,

Thank you for the rebase. Here you can see Travis build after the rebase: https://travis-ci.org/franekp/libcypher-parser/builds/307076267 and Travis build of the autotools integration attempt: https://travis-ci.org/franekp/libcypher-parser/builds/305047499.

Regarding `./configure --with-python && make all` - since I am not an Autotools expert, this is likely to be a significant amount of work. I can do it, but I need to understand the use-case: what type of users will prefer Autotools-based installation over installing from PyPI and what type of packages these commands should be building.

Since the Autotools integration might involve a lot of work and it is not entirely clear how exactly it should look like, I would very much prefer to do it in a separate PR.

I suggest that:
 - I fix things that were broken by the rebase.
 - You point out any other concerns that are fixable in a reasonable amount of time.
 - I fix them.
 - This PR is finally merged so that time is not wasted on unnecessarily rebasing and fixing things.
 - You set up Travis build on the main repo so that future PR-s are automatically tested.
 - After this, I can do smaller PR-s that are easier to review and in general more manageable.

What do you think about this plan?

Cheers,
Franek

cleishm commented:
Hey @franekp,

Most of the integration is already done in that final commit I added. I'm just unsure how the packaging using setuptools works from there.

Note that I'm not suggesting users would install pycypher using autotools - only that the appropriate python packages are built and ready for publishing. I think this would involve calling your setuptools script, or the pip utility, in the same manner that you currently do from the shell scripts in `/ci`? Basically, I'd like to replace those shell scripts with targets in the automake generated Makefiles, and integrate them enough that `make all` also builds the python packages, and `make check` runs the tests.

Ideally, that would be in place before we merge this PR into the main trunk. Happy to put it on a branch meanwhile.

Cheers,
Chris

franekp commented:
How have you been testing the code in final commit?

franekp commented:
Btw, it was integrated enough that make check was building and testing the bindings.

franekp commented:
So you want `make all` to build python packages for all Python versions both 32 and 64 bit?

cleishm commented:
The final commit is a work-in-progress. I just pushed a fix for the issue that your travis build picked up, to the branch here: https://github.com/cleishm/libcypher-parser/tree/pycypher-wip

As for the architecture - I think just building using whatever version of python that is picked up by autoconf is fine, at least for the local build/test. When it comes to building a release, it makes sense to do that using your docker environment and all the python versions available in the pypa/manylinux1_i686 image.

franekp commented on pycypher/src/setup.py:
```diff
@@ -51,7 +51,7 @@ def read_version():
     long_description=long_description,
     license='Apache License 2.0',
-    url=u'https://github.com/cleishm/libcypher-parser/tree/master/pycypher',
+    url=u'https://github.com/cleishm/libcypher-parser/tree/master',
```
This link is shown on PyPI and users looking for documentation on how to use python bindings will click it.
Please explain why you think it is better to show documentation of underlying C library to users interested in using a Python library?

cleishm replied:
I'm happy to add something about the python bindings into the top level README. The README in the subtree seemed to be more oriented at the usage of pycypher rather than at building the source tree?

franekp replied:
What is wrong with readme in subtree being oriented at usage? This is where users should be directed to from PyPI page.

cleishm replied:
If that URL is where users will find documentation of how to use pycypher, that's certainly different. In which case, I agree that it should point to https://github.com/cleishm/libcypher-parser/tree/master/pycypher/doc. Unless you want to publish a webpage? You are welcome to create a subpage on http://cleishm.github.io/libcypher-parser.

franekp replied:
I'm glad you agree with it. Regarding webpage - maybe later.

franekp commented:
Let's do the development on this PR, here at least we have travis builds (although not linked from the PR...).

franekp commented:
I will remove the reverting commit.

franekp commented:
Please address the comments.

franekp commented on .dockerignore:
```diff
@@ -1,2 +0,0 @@
-*
```
Probably a mistake?

cleishm replied:
Yeah, that file should probably remain.

franekp commented:
Oh, it seems that GitHub does not allow to comment renamed files. So: `pycypher/README.md → pycypher/doc/README.md`
1. This file contains relative links, they should be adjusted when moving the file.
2. This file was intended to be linked from PyPI so that users can find documentation easily.

cleishm commented on pycypher/src/Makefile.am:
```diff
@@ -0,0 +1,49 @@
+generated_bindings = getters.py node_types.c operators.c props.c
+
+.NOTPARALLEL: $(generated_bindings)
```
Make is often run using parallelization (by setting `MAKEFLAGS=-j4` for example).

This rule covers four output targets (getters.py, node_types.c, operators.c & props.c). If make was running in parallel, it would invoke this rule concurrently for each output.

In this case, however, your generation script builds all four output files at once. So there is no need to run in parallel as all dependencies will be fulfilled by a single invocation.

cleishm commented on pycypher/src/Makefile.am:
```diff
@@ -0,0 +1,49 @@
+	$(PYTHON) $(top_srcdir)/build-aux/pycypher/generate-bindings.py < $<
+
+pkgpyexec_LTLIBRARIES = pycypher.la
+pycypher_la_SOURCES = \
```
I don't know. Certainly AM_PATH_PYTHON (https://www.gnu.org/software/automake/manual/html_node/Python.html) suggests using these, but I'm unclear on how pip or setuptools work yet, so this was just an experiment.

cleishm replied:
BTW - I did get it to build a python module using these targets, but I couldn't get the testing to work. It seemed to fail in loading the shared library, due to a runtime linking issue (it expected the libcypher-parser.so to be already installed, rather than using the version compiled from the source tree).

franekp replied:
You probably need to adjust LD_LIBRARY_PATH

cleishm replied:
Unfortunately, that doesn't work. The bindings library has an RPATH compiled into it, based on the target install location of libcypher-parser.so. I need to figure out how to adjust that so that it can be run against the built version.

BTW - I found that the `check_pycypher.sh` was actually picking up the installed version of libcypher-parser, not the version from the source tree.

franekp replied:
So you compiled bindings using autotools and not `setup.py build_ext` ? I would suggest to just call the latter from Makefile.am, then the `.so` should not have any RPATH compiled inside it.

franekp replied:
It definitely should not have RPATH compiled in it.

cleishm replied:
I compiled both ways on my macos development system and both included RPATH into the bindings.so. I need to do some research to see how to handle that when doing trying to run tests locally.

cleishm commented on pycypher/src/setup.py:
```diff
@@ -19,12 +19,12 @@
 bindings = Extension(
```
I was trying to normalize the packaging to `<component>/(src|test)`. I think the best is to stage all the build artifacts into a pycypher directory (perhaps `.build/pycypher`, and then package.

Note that I did have it building a module, but was having trouble with the testing. That was also true when building with the setup.py.

franekp replied:
Different languages have different file structure conventions. Sure you can do hacks to structure python project the same way as c project but is it worth the time?

franekp replied:
What problems did you encounter using `setup.py` ?

cleishm commented on pycypher/src/Makefile.am:
```diff
@@ -0,0 +1,49 @@
+$(generated_bindings): $(top_srcdir)/lib/src/cypher-parser.h.in
+	$(PYTHON) $(top_srcdir)/build-aux/pycypher/generate-bindings.py < $<
+
+pkgpyexec_LTLIBRARIES = pycypher.la
```
That's a target introduced by AM_PATH_PYTHON, part of automake for building python tools/libs: https://www.gnu.org/software/automake/manual/html_node/Python.html

berggren commented:
@cleishm What is the status of this PR? Can it be merged in this state, or is there anything that needs fixing before? Please let @franekp know so he can fix any outstanding issues.

I would very much like to see this feature added :)

Thanks for all your work here!

cleishm commented:
Hi @berggren,

It's not quite ready to merge - I'm still experimenting with the best way to integrate the python package builds with the existing autotools build chain. I need to learn a bit more about how python packaging works :)

Currently, it builds and installs using autotools directly (rather than python setuptools) - but doesn't run the test suite. For that, I think I need to build the package in the local tree somehow. I think I'll switch back to using setuptools, invoked via make, which will hopefully solve that also (see https://blog.kevin-brown.com/programming/2014/09/24/combining-autotools-and-setuptools.html).

Happy to have any assistance, if this is of interest to you. Otherwise, I expect to take another look at it this weekend.

Cheers,
Chris

berggren commented:
Thanks @cleishm for the update! Sounds good, I just wanted to make sure this was being worked on as it would be a pitty if this PR was dropped. Lot's of good work went into it and having these bindings is really awesome for the project :)

Thanks for working in this!

berggren commented:
Friendly ping @cleishm any updates after the holidays? :)
Any ETA on when you can merge this?

berggren commented:
Hey @cleishm did you get a chance to look at this PR again?

cleishm commented:
@berggren Honestly, no. I've been flat out on another project. I'd love some help on this - feel free to jump in if you can!

berggren commented:
Cool, I totally understand. I'll see if I can find time to look into this. I will also reach out to Franek to see if he has some cycles over.

Thanks for the update.

cleishm commented:
If this is still valid, please feel free to reopen against the main branch.
